File: string-hard/10.正则表达式匹配.py
#
# @lc app=leetcode.cn id=10 lang=python3
#
# [10] 正则表达式匹配
#

# @lc code=start
class Solution:
    def isMatch(self, s: str, p: str) -> bool:

        @lru_cache(None)
        def recursion(i: int, j: int) -> bool:
            if j == len(p):
                return i == len(s)
            first_match = (len(s) > i) and (p[j] == s[i] or p[j] == '.')
            if len(p) > j+1 and p[j+1] == '*':
                return recursion(i, j+2) or (first_match and recursion(i+1, j))
            return first_match and recursion(i+1, j+1)

        return recursion(0, 0)

# @lc code=end

File: dynamic-programming-hard/10.正则表达式匹配.cpp
/*
 * @lc app=leetcode.cn id=10 lang=cpp
 *
 * [10] 正则表达式匹配
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    bool isMatch(string s, string p) {
        int m = s.length(), n = p.length();

        auto matches = [&](int i, int j) {
            if (i == 0)
                return false;
            else if (p[j - 1] == '.')
                return true;
            else
                return s[i - 1] == p[j - 1];
        };

        vector<vector<int>> f(m + 1, vector<int>(n + 1));
        f[0][0] = 1;
        for (int i = 0; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (p[j - 1] == '*') {
                    f[i][j] |= f[i][j - 2];
                    if (matches(i, j - 1)) {
                        f[i][j] |= f[i - 1][j];
                    }
                } else {
                    if (matches(i, j)) {
                        f[i][j] |= f[i - 1][j - 1];
                    }
                }
            }
        }
        return f[m][n];
    }
};
// @lc code=end
